src/radix_connect/mobile/client.rs

```diff
@@ -1,49 +1,174 @@
+    /// Validation consists in verifying the origin of the session.
+    new_sessions: RwLock<HashMap<SessionID, Session>>,
+}
+
+// Provisional API
+#[uniffi::export]
+impl RadixConnectMobile {
+    // RadixConnectMobile should require a NetworkAntenna and a SecureStorage from the Wallet.
+    // The internal components, such as RadixConnectRelayService will be created by the RadixConnectMobile.
+    #[uniffi::constructor]
+    pub fn new(
+        network_antenna: Arc<dyn NetworkAntenna>,
+        secure_storage: Arc<dyn SecureStorage>,
+    ) -> Self {
+        Self {
+            relay_service: RelayService::new_with_network_antenna(
+                network_antenna,
+            ),
+            secure_storage: WalletClientStorage::new(secure_storage),
+            new_sessions: RwLock::new(HashMap::new()),
+        }
+    }
+
+    #[uniffi::method]
+    pub async fn handle_linking_request(
+        &self,
+        request: RadixConnectMobileLinkRequest,
+    ) -> Result<Url> {
+        // 1. Get the handshake request from the relay service
+        let handshake_request = self
+            .relay_service
+            .get_session_handshake_request(request.session_id)
+            .await?;
+
+        // 2. Generate a new Diffie-Hellman key pair
+        let wallet_private_key = DiffieHellmanPrivateKey::generate()?;
+
+        // 3. Compute the shared secret
+        // random salt
+        let salt = hex_decode("000102030405060708090a0b0c0d0e0f").unwrap();
+        let info = hex_decode("f0f1f2f3f4f5f6f7f8f9").unwrap();
+        let encryption_key = wallet_private_key.hkdf_key_agreement(
+            &handshake_request.public_key,
+            &salt,
+            &info,
+        )?;
+
+        // 4. Create a new session
+        let session = Session::new(
+            request.session_id,
+            SessionOrigin::WebDapp(request.origin),
+            encryption_key,
+        );
+
+        {
+            self.new_sessions
+                .try_write()
+                .map(|mut new_sessions| {
+                    new_sessions.insert(request.session_id, session);
+                })
+                .unwrap();
+        }
+
+        // 5. TODO: use the actual dapp callback path
+        Ok(Url::from_str("https://example.com").unwrap())
+    }
+
+    #[uniffi::method]
+    pub async fn handle_dapp_interaction_request(
+        &self,
+        dapp_request: RadixConnectMobileDappRequest,
+    ) -> Result<RadixConnectMobileSessionRequest> {
+        let session;
+        {
+            let new_sessions = self.new_sessions.try_read().unwrap();
+
+            session = match new_sessions.get(&dapp_request.session_id) {
+                Some(session) => {
+                    // Save the session to the secure storage
+                    self.secure_storage.save_session(session.clone())?;
+                    let session_id = session.id;
+
+                    self.new_sessions
+                        .try_write()
```

Comment:
ah third place you access the `RwLock` (two writes and one read), consider using this pattern I use for `Wallet`'s `RwLock` of `Profile:

https://github.com/radixdlt/sargon/blob/develop/src/wallet/wallet.rs#L214-L222

and 
https://github.com/radixdlt/sargon/blob/develop/src/wallet/wallet.rs#L240-L253